### DISCUSSION TITLE
How to setup rust-analyzer to use cross for all the building?

### DISCUSSION THREAD

**Discussion by becothas** [cross-rs/cross#1459] [Q&A] [answered]
Hi. 
I am trying to setup cross to use together with rust-analyzer. 

To get cross to work, I had to create a modified Dockerfile, that only copies my libs and *.h files into the image. After that cross worked as expected, but rust-analyzer cannot expand procMacros and searches the target directory in my os-root. 

My current rust-analyzer configuration.
```{
    "rust-analyzer": {
        "check":{
            "features":"all",
            "overrideCommand": ["cross", "check", "--message-format=json","--target=armv7-unknown-linux-gnueabihf"],
            "targets": "armv7-unknown-linux-gnueabihf"
        },
        "cargo":{
            "buildScripts": {
                "overrideCommand": ["cross", "check", "--message-format=json","--target=armv7-unknown-linux-gnueabihf"]
            },
            "target": "armv7-unknown-linux-gnueabihf"
        }
    }
}
```
Dockerfile 
```
FROM ghcr.io/cross-rs/armv7-unknown-linux-gnueabihf:edge


COPY sysroot/lib /lib
COPY sysroot/usr /usr
```
Cross.toml 
```
[build]

default-target = "armv7-unknown-linux-gnueabihf"

[target.armv7-unknown-linux-gnueabihf]
image = "cross-env:latest"

```
Rust analyzer version rust-analyzer 1.76.0
Host Arch is x86_64 with Linux

Is there a guide to setup Rust-analyzer with cross to have all the building happen with cross, because of dependencies that are precompiled for the target platform ?
Or can somebody give me some support to get it working?

**Comment by Emilgardis** [comment]
This is more a rust-analyzer problem. We shouldnt do anything to make this not work.

**Comment by becothas** [comment, marked as answer]
Thanks for your reply. 
I discussed the issue in the rust forum [Discussion](https://users.rust-lang.org/t/how-to-setup-rust-analyzer-to-use-cross-rs/108447)

The solution was, to just start rust-analyzer in a cross container with 
`cross-util run -i --target <target> -- rust-analyzer`

**Reply by LeCyberDucky** [reply]
Could somebody please elaborate on how to do this? 

I am developing with rust-analyzer in VS Code on a Windows 10 machine. I need to use cross, since I am running into this problem https://github.com/cross-rs/cross/discussions/1462.

So, my project now inlcudes a Cross.toml file containing this: 

```
[build]
dockerfile = "dbus.Dockerfile"
default-target = "armv7-unknown-linux-gnueabihf"
```

and a dbus.Dockerfile file containing this:

```
ARG CROSS_BASE_IMAGE
FROM $CROSS_BASE_IMAGE

ARG CROSS_DEB_ARCH

RUN dpkg --add-architecture $CROSS_DEB_ARCH
RUN apt-get update && apt-get -y install libdbus-1-dev:$CROSS_DEB_ARCH
```

With that setup, I can now run `cross build` to build my project without running into my initial `libdbus` problem. 
But rust-analyzer is not aware of this fix. Accordingly, rust-analyzer runs into exactly the same problem. 

How do I need to configure VS Code and rust-analyzer to solve this? I'm not quite sure what it means to "start rust-analyzer in a cross container", since I don't start rust-analyzer manually. That is taken care of by VS Code.

Edit: I have located my rust-analyzer binary in the VS Code extensions directory. 
When I try to run `cross-util run -i --target armv7-unknown-linux-gnueabihf -- /mnt/c/Users/<USER>/.vscode/extensions/rust-lang.rust-analyzer-0.3.2002-win32-x64/server/rust-analyzer.exe`, though, I simply get

`sh: 1: /mnt/c/Users/<USER>/.vscode/extensions/rust-lang.rust-analyzer-0.3.2002-win32-x64/server/rust-analyzer.exe: not found`

**Reply by Emilgardis** [reply]
to solve that error, you'll need to expose the path with https://github.com/cross-rs/cross/blob/main/docs/config_file.md#buildenv

**Reply by bernardolansing** [reply]
It seems that `cross-util run` isn't a command anymore. Are you aware of other way to do it?

**Reply by Emilgardis** [reply]
@bernardolansing it's still available, but not in v0.2.5. you need to install from source like this:
`cargo install cross --git https://github.com/cross-rs/cross`
